packages/@lwc/state/src/index.ts

```diff
@@ -1 +1,167 @@
+  get value() {
+    return this._value;
+  }
+}
+
+class ComputedSignal<T> extends SignalBaseClass<T> {
+  private computer: Computer<unknown>;
+  private dependencies: Record<string, Signal<unknown>>;
+  private _value: T;
+  private isStale = true;
+
+  constructor(inputSignalsObj: Record<string, Signal<unknown>>, computer: Computer<unknown>) {
+    super();
+    this.computer = computer;
+    this.dependencies = inputSignalsObj;
+
+    const onUpdate = () => {
+      this.isStale = true;
+      this.notify();
+    };
+
+    for (const signal of Object.values(inputSignalsObj)) {
+      signal.subscribe(onUpdate);
+    }
+  }
+
+  private computeValue() {
+    const dependencyValues: Record<string, unknown> = {};
+    for (const [signalName, signal] of Object.entries(this.dependencies)) {
+      dependencyValues[signalName] = signal.value;
+    }
+    this.isStale = false;
+    this._value = this.computer(dependencyValues) as T;
+  }
+
+  protected override notify(): void {
+    this.isStale = true;
+    super.notify();
+  }
+
+  get value() {
+    if (this.isStale) {
+      this.computeValue();
+    }
+    return this._value;
+  }
+}
+
+const isUpdater = (signalOrUpdater: Signal<unknown> | ExposedUpdater) =>
+  typeof signalOrUpdater === 'function';
+
+const atom: MakeAtom = <T,>(initialValue: T) => new AtomSignal<T>(initialValue);
+
+const computed: MakeComputed = (inputSignalsObj, computer) =>
+  new ComputedSignal(inputSignalsObj, computer);
+
+const update: MakeUpdate = <
+  SignalSubType extends Signal<unknown>,
+  AdditionalArguments extends unknown[],
+  Updater extends (signalValues: ValuesObj, ...args: AdditionalArguments) => ValuesObj,
+  SignalsObj extends Record<string, SignalSubType>,
+  ValuesObj extends {
+    [SignalName in keyof SignalsObj]?: UnwrapSignal<SignalsObj[SignalName]>;
+  },
+>(
+  signalsToUpdate: SignalsObj,
+  userProvidedUpdaterFn: Updater,
+) => {
+  return (...uniqueArgs: AdditionalArguments) => {
+    const signalValues = {} as ValuesObj;
+
+    for (const [signalName, signal] of Object.entries(signalsToUpdate)) {
+      signalValues[signalName as keyof ValuesObj] = signal.value as ValuesObj[keyof ValuesObj];
+    }
+
+    const newValues = userProvidedUpdaterFn(signalValues, ...uniqueArgs);
+
+    for (const [atomName, newValue] of Object.entries(newValues)) {
+      signalsToUpdate[atomName][atomSetter](newValue);
+    }
+  };
+};
+
+export const defineState: DefineState = (defineStateCallback) => {
+  return (...args) => {
+    class StateManagerSignal<OuterStateShape> extends SignalBaseClass<OuterStateShape> {
+      private internalStateShape: Record<string, Signal<unknown> | ExposedUpdater>;
+      private _value: OuterStateShape;
+      private isStale = true;
+
+      constructor() {
+        super();
+
+        // @ts-ignore TODO: W-16769884
+        const fromContext: MakeContextHook = () => {};
+
+        this.internalStateShape = defineStateCallback(atom, computed, update, fromContext)(...args);
```

Comment:
For a follow-up work item:

We probably want to validate `internalStateShape` at this point. For example, we would want to confirm that `Object.values(this.internalStateShape)` consists only of atoms, computed, and updaters. Including other signals here might result in subtle bugs, so we should explicitly disallow that. Additionally, someone might try including something that is not signal-like and also not an updater - that should also be disallowed, as the `.subscribe` code below might break in that case.